rfcs/0005-multi-blockchain-support.md

```diff
@@ -0,0 +1,487 @@
+
+// Types available now:
+//
+// - ethereum.Block
+// - ethereum.Event
+// - ethereum.Contract
+// - ethereum.Call
+// - ethereum.Value
+//
+// - substrate.Block
+// - ...
+```
+
+Most, but not all of these, are internal to code generation. The same goes for
+low-level types such as Ethereum's `uint256` or `bytes32` type. These are
+already sufficiently abstracted behind `Bytes`, `BigInt`, `BigDecimal` and the
+code generated for Ethereum tuples.
+
+Additional types may be necessary for other chains, but it expected that the
+types implemented today are sufficient to start with.
+
+> **Note:** An alternative to the above could be to split chain modules up into
+> their own files or event packages, allowing them to be imported with either
+> ```js
+> import { ... } from '@graphprotocol/graph-ts/ethereum'
+> import { ... } from '@graphprotocol/graph-ts/substrate'
+> ```
+> or
+> ```js
+> import { ... } from '@graphprotocol/ethereum'
+> import { ... } from '@graphprotocol/substrate'
+> ```
+
+Each blockchain integration must include such a module as well as a WASM host
+exports implementation to back this module and type conversions from/to
+AssemblyScript for low-level types.
+
+> **Note:** Integrating a blockchain initially requires modifications across
+> Graph Node, Graph CLI and Graph TS. A more convenient extension concept can
+> still be developed later.
+>
+> - Graph Node: An alternative could be optional dependencies behind feature flags, to pull in
+>   support for specific chains at build time.
+>
+> - Graph CLI: Gluegun, used for Graph CLI, includes an extension framework that
+>   would allow to inject code generation and type conversion support from
+>   separate NPM packages.
+>
+> - Graph TS: Separate NPM packages for chain-specific mapping APIs and types as
+>   suggested in the previous note would make extending Graph TS easy.
+
+### Subgraph Indexing
+
+This section discusses how the subgraph indexing components and data structures will change, with
+some being reused and others becoming chain-specific and therefore abstracted by Rust traits. [This
+diagram](https://whimsical.com/multi-blockchain-13ZinTmYAUn5YknGjcZc4e) gives an overview of the
+desired architecture for a running subgraph. The arrows describe the flow of data.
+
+Data enters the system by being requested from the chain
+itself, for example through JSON-RPC calls. The __adapters__ are the components responsible for
+interacting directly with the chain. They are broken up into specialized traits for each component
+that requires an adapter, such as the `IngestorAdapter` and `TriggersAdapter`. For performance, the
+adapter may cache and index into the local database the data pulled from the chain. Each chain will
+have its own DB schema (aka namespace) under which adapters may create any necessary custom DB
+tables.
+
+The __block ingestor__ and the __chain store__ will be reused across chains. There will be one
```

Comment:
There's an inherent chain-specific dependency between the adapter and the block stream, which communicate with each other through data in the database: the adapter writes data in a chain-specific format into the chain store, and the block stream reads it. A lot of that can probably be handled inside the implementation of the chain itself, but the current `ChainStore` trait will at least need to be parametrized on the type of the block (currently `EthereumBlock`) Ideally though, we could remove those methods from the public chain interfaces and keep the notion of what a block looks like within the chain implementation.

Reply:
I've noted here that these components will need to be reworked to be made generic and reusable.